### ISSUE TITLE
[BUG] Haste & DoTs Not Properly Implemented

### ISSUE THREAD

**Issue by ssabbyccatt** [JamminL/cata-classic-bugs#872]
#### 1.1. Description
Haste on DoTs is currently implemented as it is in WOTLK, wherein the number of ticks never changes as you gain haste. You have a set number of ticks, and haste simply reduces the time between those ticks. We have confirmed this with Balance Druid Insect Swarm & Moonfire, along with Spriest Shadow Word: Pain, so suspect it to involve all classes with DoTs, and perhaps HoTs?
  #### 1.2. How to Reproduce
           1. Cast any DoT without haste effects, observe # of ticks
           2. Cast any DoT accompanied by haste effects. In our tests, we used a troll boomkin. Moonkin form (5% haste), nature's grace (15% haste), berserking (20%) haste. Again, observe # of ticks, which remains the same.
  #### 1.3  Source Material
- Video: https://cdn.discordapp.com/attachments/1154479397588435085/1218427095944925234/World_Of_Warcraft_2024.03.16_-_01.12.29.01.mp4?ex=66079fa8&is=65f52aa8&hm=1d40325a49947a6ddbe2f375565ea2c488ed58a8e6200853264a05d5996de83a&
- Video: https://cdn.discordapp.com/attachments/1154479397588435085/1218431003325108244/World_Of_Warcraft_2024.03.16_-_01.26.16.02.mp4?ex=6607a34c&is=65f52e4c&hm=07f9a20dd2546464939db12f19f7f2ac242d6d0c45bca355ca17bf888e2ac4c7&
- Moonkin form + nature's grace + berkserking: https://cdn.discordapp.com/attachments/1154479397588435085/1218432055625515038/Screenshot_2024-03-16_013315.png?ex=6607a447&is=65f52f47&hm=2dd6efd3848ebd2e0a93a278a35e59130664f7bf95ddf7cc1a33a2c80556490a&
- Only moonkin form: https://cdn.discordapp.com/attachments/1154479397588435085/1218437744913809438/Screenshot_2024-03-16_015628.png?ex=6607a993&is=65f53493&hm=39c0862808bd8dd3b23481c30baf4dfa5ee6437e5f183e69558e169bf3f7444c&
- Spriest with no extra ticks at 50% cast speed: https://cdn.discordapp.com/attachments/1154479397588435085/1218436968254410752/image.png?ex=6607a8da&is=65f533da&hm=54d0648455f0f4d049269ae3da559bcf3d0353898dc0524b84689cdf6f491cfa&
#### 2.2 Source Material
Included are a number of articles that explain how haste & dots in Cata should work. The first goes in-depth into the dot/haste functionality, and the rest show that Moonkin dots specifically should also function in this manner:

- Haste & DoTs in Cata: https://cynwise.wordpress.com/2010/10/29/how-warlock-dots-work-in-cataclysm/ 
- Moonkin Cata guide from 2010, "Haste will affect...the frequency of your Moonfire, Sunfire, Insect Swarm and Hurricane Ticks": https://www.mmo-champion.com/threads/874888-The-Moonkin-Guide-Cataclysm-Edition
- Additional sources: http://graymatterwow.blogspot.com/2011/02/mailbag-haste-breakpoints.html 
- https://www.engadget.com/2010-07-30-shifting-perspectives-balance-dots-in-cataclysm.html?guccounter=1&guce_referrer=aHR0cHM6Ly93d3cuZ29vZ2xlLmNvbS8&guce_referrer_sig=AQAAADSLo1MWGy5RXwLhFmW6HsE7-SbI1BBphIt5yoTYa2n_wNK8NNshzG_Nyfag4DH-R03sKJOvsNwgAFk0I32Tj7X7rR5cx_P-Ys-dO1AbhVF7X-HRdukTaqdD6oLAUwENAp71vnHDubfO4tBIvYC2m_dH2ISDeNUyAVKuF5-d3i99

**Comment by ssabbyccatt** [comment]
Update: after most recent patch, haste seems to be mostly(?) working correctly, though I've heard it does not with at least Spriest VT. No dots on any class can crit ATM.

**Comment by TamerLee** [comment]
> Update: after most recent patch, haste seems to be mostly(?) working correctly, though I've heard it does not with at least Spriest VT. No dots on any class can crit ATM.

As of today, haste is reducing the duration of dots for warlocks. Corruption has a base duration of 18 seconds, but haste is reducing this duration:
![image](https://github.com/JamminL/cata-classic-bugs/assets/164273628/00c61b29-c2bf-4b74-84db-92a5fda4f3e8)

Per source link in OP, haste should not decrease the duration of dots:
![corruption-haste-comparison-1](https://github.com/JamminL/cata-classic-bugs/assets/164273628/65380b82-eec8-4865-b581-90d1aaaeb040)

**Comment by Lillyton** [comment]
> Additionally, refreshing a dot just before the final tick does not allow the final tick to occur.

Correct. This is how it worked in Cata. The pandemic mechanic was not in Cata.

**Comment by TamerLee** [comment]
> > Additionally, refreshing a dot just before the final tick does not allow the final tick to occur.
> 
> Correct, not a bug. This is how it worked in Cata. The pandemic mechanic was not in Cata.

I'm not referring to the pandemic mechanic. It works a bit different than both pandemic and how dots worked in Wrath. 

DoT clipping refers to the action of refreshing a DoT before its final tick has occurred. If you refresh a DoT while it had 2 or more ticks left, all ticks except for the final one will be "wasted". The final tick will be added to the new duration of the DoT (which will, therefore, be extended to include it). Furthermore, the final tick of the previous DoT, when becoming part of the new DoT, will be modified to your current statistics, and not the statistics of the old DoT.

What this means, in simpler terms, is that:

-If you refresh a strong application of a DoT with a weaker one, between its penultimate and last ticks, the final tick will become weak.
-If you refresh a weak application of a DoT with a stronger one, between its penultimate and last ticks, the final tick will become strong.

This behavior is documented here: https://cynwise.wordpress.com/2010/10/29/how-warlock-dots-work-in-cataclysm/

What happens when you refresh a dot just before the final tick in wrath:
![corruption-clipping-dot-2](https://github.com/JamminL/cata-classic-bugs/assets/164273628/5e986ddb-5092-41d5-8d32-df745c0e3bab)

What happens when you refresh a dot just before the final tick in cataclysm:
![corruption-good-refresh-1-3](https://github.com/JamminL/cata-classic-bugs/assets/164273628/33ac13c2-6d59-4a80-8052-bb8d07ca9deb)

**Comment by ssabbyccatt** [comment]
As of today's updates, all dots are still not critting and do not have the final-tick implementation they should have.

**Comment by InDebt** [comment]
I want to add here that it should not only be the last tick. If we refresh a DoT the next tick will carry over as stated in the referenced guide.

> This example shows refreshing the DoT after the third tick.  The fourth tick still happens – it doesn’t get clipped – but the duration doesn’t stack up to a full 36 seconds.

This needs to be tested as well as it will impact refreshing DoTs during movement phases.

And I guess since this is about DoTs in generall and the new Cata mechanics is should be noted that the tick that gets carried over should already be subject to the new snapshots.

https://youtu.be/hbmtLM3xBsI?si=Bw8xVFPahYa4b1rm&t=42 That should be Lightweave and Power Torrent running as a buff. DP is rebuffed before the last tick and the tick that get carried over ticks for 2.4k instead of the previous 2.2k.

**Comment by ssabbyccatt** [comment]
As of 4/18 beta update, Moonfire and Insect Swarm are still not critting, and still lack the extra-tick implementation.

**Comment by ssabbyccatt** [comment]
> As of 4/18 beta update, Moonfire/Sunfire and Insect Swarm are still not critting, and still lack the extra-tick implementation. It sounds like many other DoTs are critting, but Boomie DoTs are not.

This is all still bugged as of the 4/22 beta update.

**Comment by ssabbyccatt** [comment]
The dot tick implementation is working for boomkin on the live prepatch servers. @TamerLee @InDebt Can yall check whatever non-Boomkin class I assume you're playing and see if it is working now for you as well? If it is I'll close this ticket

**Comment by InDebt** [comment]
Cheers. So DoTs were able to crit for a while and the haste calculations are correct as well. Only thing missing currently is the carry over of the next tick if you rebuff the dot before expiration. Does that work for balance?

**Comment by ssabbyccatt** [comment]
> Cheers. So DoTs were able to crit for a while and the haste calculations are correct as well. Only thing missing currently is the carry over of the next tick if you rebuff the dot before expiration. Does that work for balance?

As of this moment, for NA (I think EU isn't on same patch yet), balance dots are working in terms of haste, crit, and the carrying over of the tick if you rebuff before expiration. Tested multiple times this morning, you can see it in the image below (the 2nd MF cast occurred at approx 1s left on the initial MF duration, and 24 ticks was the expected result if it is working properly). So, at least for balance, things seem to be good to go on DoTs. 
![Screenshot 2024-05-01 110703](https://github.com/JamminL/cata-classic-bugs/assets/99249505/4c7c88d7-22d5-4cff-8286-118eb509a887)

**Comment by InDebt** [comment]
So tested right now. Build `54500` which should be the same as NA.
Tick timer is definetly still reset for shadow spells as seen in the screenshot.

Refreshing the dot will delay the tick timer which is not show it should work,
![grafik](https://github.com/JamminL/cata-classic-bugs/assets/1821006/4ad19c2e-0314-4d17-b664-89780b2fe5f3)

**Comment by ssabbyccatt** [comment]
It seems like there actually is still an issue with boomie tick cadence, you can see it in this log (go to Udderrless) - the tick cadence is interrupted when recasting MF. Without timestamps it appeared to be working, but when you actually log it and look at the timestamps it is not. https://classic.warcraftlogs.com/reports/ZJdhvxLfQ4V2W6Kn#fight=1&source=7&type=damage-done&ability=8921&view=events

**Comment by InDebt** [comment]
Alright with the latest build 54501 this seems to be resolved

**Comment by ssabbyccatt** [comment]
Still not resolved for boomie

**Comment by ssabbyccatt** [comment]
Closing as Blizz has specifically addressed this https://www.wowhead.com/cata/news/may-3rd-cataclysm-classic-hotfixes-bloodlust-fixed-dots-working-correctly-339202?webhook
